### DISCUSSION TITLE
Script compiler for MQMGateway

### DISCUSSION THREAD

**Discussion by fencepost-error** [BlackZork/mqmgateway#76] [Show and tell]
I've just updated my [script compiler for MQMGateway](https://github.com/fencepost-error/Templates-for-MQMGateway)  (and Home Assistant), this allows generation of the required YAML to run everything with minimal effort, for example the following sets up nine modbus devices producing around 100kB of YAML from only one line of code per device:

```
compile_header

compile_template eastron grid indoors 10

compile_template waveshare_io waveshare_io indoors 20
compile_template pulse_count water indoors 21

compile_template temp_humid basement basement 10

compile_template temp_humid front_door outdoors 10
compile_template temp_humid back_door outdoors 11
compile_template co2 co2 outdoors 13
compile_template pm25 pm25 outdoors 14

compile_template water_level water_tank outdoors 20
```
Quoting from the readme, it:
* Applies a single template file to as many devices as you specify.
* Translates device DNS names to their IP addresses.* 
* Performs checks for common errors in config files.
* Generates the required config.yaml and applies it to modmqttd.
* Generates the YAML required for use with Home Assistant.
* Generates corresponding JSON for use in Home Assistant dashboards.

This means that the entire process of creating the YAML config file and applying it to modmqttd requires only a single command, and the next step to Home Assistant just requires adding the YAML to the appropriate HA config file and the corresponding JSON to the desktop you want to display the sensors on.

It's available under the same license as MQMGateway.

**Comment by BlackZork** [comment]
100kB YAML? Is it a mqmgateway config file or homeassistant? Or both?

I have 3 modbus networks and 13 slaves in total and my config file is 25kB.

**Reply by fencepost-error** [reply]
It's the combined mqmgateway + HA YAML + HA dashboard config.  Also when you've got x devices * y registers in each you get a rapid explosion of entries, which was the main motivation for creating the script, any change propagates across so many entries that it made sense to automate it.

Another issue is that the script compiler to automate setup and availability of cheap Aliexpress devices is a dangerous combination :-).

**Comment by fencepost-error** [comment]
I've just updated the script compiler to fix a bug where commands inside comments weren't ignored, and added a few more templates for devices, supported ones are now:

- CDEbyte digital/analog I/O devices.
- Eastron SDM 2x0 power meters.
- Generic CO2 sensors, e.g. ComWinTop.
- Generic PM2.5/PM10 sensors, e.g. ComWinTop.
- Generic pressure-based water level sensors, e.g. Seeed Studio, QuiDian.
- Generic temperature/humidity sensors, e.g. ComWinTop.
- MIA HVAC systems.
- Modbus gateway boards for DS18B20 temperature sensors, e.g. eletechsup.
- Waveshare I/O modules.
- Wayjun pulse counter modules.

**Comment by fencepost-error** [comment]
Another update, this one removes the need to hand-edit the generated YAML for some special cases.  Now it's all automated, generating the YAML config and updating MQMGateway with it is a single operation.
